src/dataregistry/registrar_util.py

```diff
@@ -207,3 +209,43 @@ def _name_from_relpath(relative_path):
         name = base
 
     return name
+
+def _copy_data(dataset_organization, source, dest):
+    """
+    Copy data from one location to another (for ingesting directories and files
+    into the `root_dir` shared space.
+
+    Note prior to this, in `_handle_data`, it has already been check that
+    `source` exists, so we do not have to check again.
+
+    Parameters
+    ----------
+    dataset_organization : str
+        The dataset organization, either "file" or "directory"
+    source : str
+        Path of source file or directory
+    dest : str
+        Destination we are copying to
+    """
+
+    # Copy a single file
+    if dataset_organization == "file":
+        # Create any intervening directories
+        os.makedirs(os.path.dirname(dest), exist_ok=True)
+
+        # Copy the file
+        copyfile(source, dest)
+
+    # Copy a single directory (and subdirectories)
+    elif dataset_organization == "directory":
+        # If the directory already exists, need to delete it first
+        if os.path.isdir(dest):
+            rmtree(dest)
```

Comment:
If anything goes wrong with the subsequent copy, the old dataset will be gone and it won't be possible to recover the old state.   It would be safer to `mv` the old dataset somewhere (e.g. to something with a  path similar to the true destination, but perhaps starting with a `.`  or with strange characters - emacs uses `#` at the beginning and end of the filename - strategically placed), copy in the new data, then if that succeeds do rmtree.  

If we're feeling really paranoid, the rmtree shouldn't happen until after the db row is successfully created.  And for datasets which are just files we should also do just a `mv` initially to a similar name, then `rm` after the database entry is made.

Reply:
I've changed it as suggested.

When the destination exists, mv it to a backup, copy the data, then delete the backup if successful. If the `try` fails, mv the backup back to the original.

If its an individual file being ingested, a checksum is also performed.

I think putting the entire database entry in a try would be a bit cumbersome, to check if the entry is also sucessful before deleting the backup.